### PR TITLE
forward packets by prioritization in desc order

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

from #23211: `keep filling account buckets and forward each bucket to the next leader to avoid starving accounts on forwards. As the account buckets are filled with CUs, forwards should start sending txs from multiple accounts, not just the one with the highest fees. Naive implementation would forward highest paid TXs first, which are likely to be only for a single hot account. Otherwise a single congested market with a ton of demand will saturate the forward queues.`

#### Summary of Changes
- using requested_cu from `compute_budget` for transaction cost (instead of using estimated cost);
- add `forward_packet_batches_by_account` to group transactions to "account buckets" in fee priority
- forward block of transactions (grouped by account buckets)

Fixes

### CODE REVIEW

**Comment by bji** [comment]
What is the purpose of this?

- Transactions are going to be sent to the leader directly.  It makes little to no sense for clients to send tx to anyone other than the leader, because that just introduces an unnecessary hop, unless the client is trying to treat the validator as a "poor man's RPC node" that knows the leader schedule and will forward tx onto the leader given that.  But are clients sophisticated and determined enough to spam the network really so unsophisticated as to not be able to track the leader schedule themselves and hit the leader directly?

- If my assertion is true, that clients are always trying to hit the leader, not some validator distant from the leader, then the number of validators who are *not* the leader who would receive transactions is a small set - probably the validator before and after the leader, hit with stray tx delivered by clients who "missed" the leader.

- Is there any purpose in sorting by validator sources when that's likely to be a set of only two or three validators, and if the client who is sending tx only hit them by 'mistake'?  It's not like you can incentivize clients to "aim better", they are almost certainly already aiming the best they can.  Or more likely, always sending their tx to validator before the leader, the validator after the leader, and the leader itself, in order to maximize their chance of hitting the actual leader.

What *would* make sense is:

- Leaders prioritize incoming transactions based on the stake weight of the node that delivered them.  This then makes it harder for clients to hit leaders directly without skin in the game of being staked more than others trying to compete to get tx to the leader.

So TL;DR:

Should prioritize on the inbound queue of the receiving validator, not the outbound queue of the sending validator.

**Comment by bji** [comment]
(note that individual validators can establish trust relationships with specific clients - especially RPC nodes - to give them some agreed-upon "stake weight equivalent" priority)

**Comment by taozhu-chicago** [comment]
this is for #23211 , especially this section:
```
keep filling account buckets and forward each bucket to the next leader to avoid starving accounts on forwards. As the account buckets are filled with CUs, forwards should start sending txs from multiple accounts, not just the one with the highest fees. Naive implementation would forward highest paid TXs first, which are likely to be only for a single hot account. Otherwise a single congested market with a ton of demand will saturate the forward queues.
```

**Comment by aeyakovenko** [comment]
@bji there is no way to guarantee that tx arrives at the node while it's still the leader.

File: core/src/forward_packet_batches_by_accounts.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        rc::Rc,
+        sync::{Arc, RwLock},
+    },
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits
+const DEFAULT_LIMIT_RATIO: u32 = 2;
+/// this number divided by`DEFAULT_LIMIT_RATIO` is the total blocks to forward
+/// means default 4 blocks worth packets will be forwarded
+const DEFAULT_NUMBER_OF_BATCHES: u32 = 8;
```
Comment (taozhu-chicago) [inline core/src/forward_packet_batches_by_accounts.rs:21]: Probably should have a much large value for this. The reason is transactions without `compute_budget` ix have 200K default units (which puts their priority lower). 120 such transaction fill up half-block (120 * 200K = 24mil units). Maybe set it to `100` to forward about 12K transactions without `compute_budget`. Thoughts?

File: core/src/forward_packet_batches_by_accounts.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        rc::Rc,
+        sync::{Arc, RwLock},
+    },
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits
+const DEFAULT_LIMIT_RATIO: u32 = 2;
```
Comment (carllin) [inline core/src/forward_packet_batches_by_accounts.rs:18]: An explanation as to why we want to less than the full tracker limit would be helpful here

File: core/src/unprocessed_packet_batches.rs [hunk 3]
```diff
@@ -245,12 +250,43 @@ impl UnprocessedPacketBatches {
...
+                            immutable_packet.message_hash()
+                        );
+                    }
+                    Entry::Occupied(mut occupied_entry) => {
+                        f(occupied_entry.get_mut());
+                        immutable_packet
+                    }
+                }
+            })
+            .collect();
```
[the remark this answers is not in the document]
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:279]: Another thought, I feel like the `fill_batches` here: https://github.com/solana-labs/solana/blob/7dc554bbe398eac0e7e8094ffca48c410df6e6f2/core/src/forward_packet_batches_by_accounts.rs#L154 will use very few of these transactions.


Like if if had 250k packets in this buffer, we might only need 5k of them to fill all the buckets. We can probably leverage this to avoid cloning the entire buffer by popping only the ones we need, and then readding them back to the main data structure later

[the remark this answers is not in the document]
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs:279]: Should try benching this with varying heap sizes to see what the perf looks like comparatively

File: core/src/banking_stage.rs [hunk 4]
```diff
@@ -1002,43 +995,81 @@ impl BankingStage {
...
-            cluster_info,
-            poh_recorder,
-            forwardable_packets,
-            data_budget,
-            banking_stage_stats,
+        let mut forward_packet_batches_by_accounts =
+            ForwardPacketBatchesByAccounts::new_with_default_batch_limits(bank_forks);
+        let forwardable_packets_count = Self::filter_valid_packets_for_forwarding(
+            buffered_packet_batches,
+            &mut forward_packet_batches_by_accounts,
```
Comment (carllin) [inline core/src/banking_stage.rs:1002]: https://github.com/solana-labs/solana/blob/7dc554bbe398eac0e7e8094ffca48c410df6e6f2/core/src/banking_stage.rs#L520-L521 We also have a forward limit of about 10,000 packets a second, so we could factor that into the bucketing as well.

If the total packets in the buckets exceed this, we can just exit and not process the rest of the buffers

File: core/src/forward_packet_batches_by_accounts.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    }
+
+    pub fn fill(&mut self, packet: Rc<ImmutableDeserializedPacket>) {
+        // do not forward packet that cannot be sanitized
+        if let Some(sanitized_transaction) =
+            unprocessed_packet_batches::transaction_from_deserialized_packet(
+                &packet,
+                &self.current_bank.feature_set,
+                self.current_bank.vote_only_bank(),
+                self.current_bank.as_ref(),
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/forward_packet_batches_by_accounts.rs:125]: Good call! Agree it is redundant to filter packets at end of slot for forwarding if forwarder is going to do the same. Removing this step simplifies banking_stage a bit more in commit  5af65401c2c9c814bd9e204e8f588099cdbb33e3

File: core/src/banking_stage.rs (hunk 4 above)
Comment (taozhu-chicago) [reply core/src/banking_stage.rs]: This is a good idea. Think can set the forward batches capacity to be inline with this limit, come to same effect of this comment https://github.com/solana-labs/solana/pull/25406#discussion_r892981705.

File: core/src/unprocessed_packet_batches.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/unprocessed_packet_batches.rs]: Did bunch of benching tests on `bankign_stage::filter_valid_packets_for_forwarding`, which includes clone/drain_desc MinMaxHeap, and sorting into buckets. With 10K packets/sec forwarding limitation in place, tested buffer of 10k forwardable packets, and 64k forwardable packets, found in both case the `filter_valid_packets_for_forwarding` took about same time (41ms, 39ms respectively). And the differences between drain_desc/collect and clone/drain_desc are also insignificant. I kept the clone/drain impl anyway, because it looks cleaner.

[the remark this answers is not in the document]
Comment (carllin) [reply core/src/unprocessed_packet_batches.rs]: Thanks for running that test, was this tested with the default 100 buckets?

File: core/src/banking_stage.rs (hunk 4 above)
Comment (carllin) [reply core/src/banking_stage.rs]: Based on that comment, couldn't it also overshoot the `12k` by a lot if a lot of the transactions are low compute unit transactions?

File: core/src/forward_packet_batches_by_accounts.rs [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits, as smaller batch
+/// allows better granularity in interating through buffer, and faster forwarding.
+const DEFAULT_LIMIT_RATIO: u32 = 2;
+/// this number divided by`DEFAULT_LIMIT_RATIO` is the total blocks to forward.
+/// To accommodate transactions without `compute_budget` instruction, which will
+/// have default 200_000 compute units, it has 100 batches as default to forward
+/// up to 12_000 such transaction. (120 such transactions fill up a batch, 100
+/// batches allows 12_000 transactions)
```
[the remark this answers is not in the document]
Comment (carllin) [reply core/src/forward_packet_batches_by_accounts.rs:20]: conversation moved here to original discussion: https://github.com/solana-labs/solana/pull/25406#discussion_r912283531

File: core/src/banking_stage.rs [hunk 7]
```diff
@@ -1988,54 +1949,6 @@ impl BankingStage {
         process_transactions_summary
     }
 
-    // Returns the number of packets that were filtered out for
-    // no longer being valid (could be too old, a duplicate of something
-    // already processed, etc.)
```
Comment (carllin) [inline core/src/banking_stage.rs:1993]: Reading this description, I was looking for where in the function we filter out transactions that are no longer valid `could be too old, a duplicate of something already processed, etc.)`

It looks like I accidentally deleted this logic in my PR:
https://github.com/solana-labs/solana/commit/870ac80b7948f22dca0eceea0061b1f74c64f56b#diff-bb8464c2f02c8863fb60e6e29f023f3e6da93a93fbb9199a70a44c9c1b41c7f3L1897-L1902

This should probably probably be added back in? This would mean the new forwarding bucket logic would have to make these sanity checks as well if we deleted it here

File: core/src/forward_packet_batches_by_accounts.rs [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        block_cost_limits,
+        cost_tracker::{CostTracker, CostTrackerError},
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::{rc::Rc, sync::Arc},
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits, as smaller batch
+/// allows better granularity in interating through buffer, and faster forwarding.
+const DEFAULT_LIMIT_RATIO: u32 = 2;
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/forward_packet_batches_by_accounts.rs:15]: done

File: core/src/forward_packet_batches_by_accounts.rs [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        bank::Bank,
+        block_cost_limits,
+        cost_tracker::{CostTracker, CostTrackerError},
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::{rc::Rc, sync::Arc},
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits, as smaller batch
+/// allows better granularity in interating through buffer, and faster forwarding.
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/forward_packet_batches_by_accounts.rs:14]: > Why does this allow for faster forwarding? Isn't the number of transactions being included roughly the same?

Good point. Thought smaller batches would go through forwarding pipeline smoother, has less chance to hit line rate limit. But think again, it doesn't really matter. It actually adds overhead of checking data_budget, and metrics. Tho not in significant way. Removed this comment to avoid confusion.

File: core/src/banking_stage.rs (hunk 4 above)
Comment (taozhu-chicago) [reply core/src/banking_stage.rs]: convo moved.

File: core/src/forward_packet_batches_by_accounts.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/forward_packet_batches_by_accounts.rs:20]: Yea, basically forwarding buckets have capacity for _at least_ 12K transactions, it could batch much more low cost transactions. This const defines _capacity_ of forwarding buckets. Currently all un-forwarded transactions in buffer are sorted into account buckets, but number of transactions being forwarded are limited by data_budget (same as before).

In the scenario that there are 20K un-forwarded transactions in buffer, knowing there is 10K/sec limit for threads, in order to avoid sorting all 20K txs, I think can put logic in `banking_stage::filter_valid_packets_for_forwarding` when buffer is iterated. I also think this can be a separate PR as enhancement. wdyt?

[the remark this answers is not in the document]
Comment (carllin) [reply core/src/forward_packet_batches_by_accounts.rs:21]: yeah i think `filter_valid_packets_for_forwarding` is a good place to truncate the batch. Separate PR sounds good to me

**Review by carllin — APPROVED** [review]

File: core/src/forward_packet_batches_by_accounts.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply core/src/forward_packet_batches_by_accounts.rs:21]: #26433

File: core/src/forward_packet_batches_by_accounts.rs [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    std::{rc::Rc, sync::Arc},
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits, as smaller batch
+/// allows better granularity in composing forwarding transactions; e.g.,
+/// transactions in each batch are potentially more evenly distributed across accounts.
+const FORWARDED_BLOCK_COMPUTE_RATIO: u32 = 2;
+/// this number divided by`FORWARDED_BLOCK_COMPUTE_RATIO` is the total blocks to forward.
+/// To accommodate transactions without `compute_budget` instruction, which will
+/// have default 200_000 compute units, it has 100 batches as default to forward
```
Comment (t-nelson) [inline core/src/forward_packet_batches_by_accounts.rs:19]: this 200k default assumption isn't essential, right? (Doesn't seem to be from these changes).  we're about to activate the feature that makes the default units `min(200k * ix_count, 1.4M)`
